bin/src/_main.rs

```diff
@@ -390,12 +381,23 @@ pub async fn _main(
             }
             _ => false,
         },
-        |(watched_dirs, rules, lookback, offsets, fo_state_handles)| {
+        |(watched_dirs, rules, lookback, offset_state, fo_state_handles)| {
+            let mut initial_offsets: Option<HashMap<FileId, SpanVec>> = None;
+
+            if let Some(offset_state) = offset_state {
+                match offset_state.offsets() {
+                    Ok(offsets) => {
+                        initial_offsets = Some(offsets.into_iter().map(|fo| (fo.key, fo.offsets)).collect());
+                    },
+                    Err(e) => warn!("couldn't retrieve offsets from agent state, {:?}", e),
+                }
+            }
+
             let tailer = tail::Tailer::new(
                 watched_dirs.clone(),
                 rules.clone(),
                 lookback.clone(),
-                offsets.clone(),
+                initial_offsets,
```

Comment:
@jakedipity 
does this mean that instead of reusing existing in-memory state we are querying fresh "initial_offsets" from database now?
Is it possible that some offset in DB may get updated after new Tailer created?

Reply:
From my understanding of how the restarting tailer stream works, it only polls the current active tailer. So if the tailer is replaced, no more events are polled from the old tailer.

I think there's still a chance lines could be somewhere in the middleware or client or retries, but that seems like a very minimal chance.

Initial offsets were never pulled from memory. They've always been calculated from the DB, but were only being done once on initial start.